ufl/duals.py

```diff
@@ -0,0 +1,23 @@
+# -*- coding: utf-8 -*-
+"Predicates for recognising duals"
+
+# Copyright (C) 2015-2016 Martin Sandve Alnæs
+#
+# This file is part of UFL (https://www.fenicsproject.org)
+#
+# SPDX-License-Identifier:    LGPL-3.0-or-later
+#
+# Modified by India Marsden, 2021
+
+# from ufl.log import error
+# from ufl.functionspace import FunctionSpace, DualSpace, MixedFunctionSpace
+
+
+def is_primal(object):
+    """ Determines if the object belongs to a primal space"""
+    return hasattr(object, '_primal') and object._primal
+
+
+def is_dual(object):
+    """ Determines if the object belongs to a dual space"""
+    return hasattr(object, '_dual') and object._dual
```

Comment:
Counterexample to law of the excluded middle :). I notice that all objects that advertise `object._dual` or `object._primal` also provide the other. It's kind of unsafe to dispatch on the result of these functions for objects that don't advertise these properties because in that case we have `(is_primal(obj) or is_dual(obj)) == False`. So should these just do `return object._primal` and `return object._dual` ?

Reply:
OK, this is a valid use case.